### PR TITLE
Improved conditional CSS rendering

### LINKED ISSUE #141: Order of media queries
### Describe the bug

Sometimes media queries appear to be in the wrong order in the CSS:

image

Here are the declarations (I'd expect same order as when created?):

image

What I find odd about this is that changing the media queries to remove the `screen and` part sometimes appears to fix it.

image

But then changing other declarations may cause it to return. Perhaps some funny shenanigans are happening with order when objects are merged to cloned? 

I'm having trouble isolating a small consistent reproduction case although I can reproduce reliably in my project.

### System Info

I'm using vanilla, vanilla extract here with no other abstractions apart from a few variables for tokens. I'm working in a mono repo with a Gatsby site where my components are stored in a different package.

Output of `npx envinfo --system --npmPackages @vanilla-extract/css,@vanilla-extract/webpack-plugin,@vanilla-extract/esbuild-plugin,@vanilla-extract/vite-plugin,@vanilla-extract/sprinkles,webpack,esbuild,vite --binaries --browsers`:

```node
  System:
    OS: Linux 4.4 Ubuntu 20.04 LTS (Focal Fossa)
    CPU: (12) x64 AMD Ryzen 5 3600 6-Core Processor
    Memory: 4.84 GB / 15.95 GB
    Container: Yes
    Shell: 5.8 - /usr/bin/zsh
  Binaries:
    Node: 12.16.3 - ~/.nvm/versions/node/v12.16.3/bin/node
    Yarn: 1.22.10 - ~/.nvm/versions/node/v12.16.3/bin/yarn
    npm: 6.14.4 - ~/.nvm/versions/node/v12.16.3/bin/npm
```

### PR DESCRIPTION
**Pull request by mattcompiles**
Resolves #141

### CODE REVIEW

File: packages/css/src/conditionalRulesets.ts [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+    return true;
+  }
+
+  /**
+   * Merge another ConditionalRuleset into this one
+   *
+   * @returns true if successful, false if the ruleset is incompatible
+   */
+  merge(incomingRuleset: ConditionalRuleset) {
```
Comment (mattcompiles) [inline packages/css/src/conditionalRulesets.ts:119]: `mergeIfCompatible`

File: packages/css/src/conditionalRulesets.ts [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+type ConditionInfo = {
+  condition: string;
+  rules: Array<Rule>;
+  children: ConditionalRuleset;
+};
+
+type Ruleset = Array<ConditionInfo>;
+
+export class ConditionalRuleset {
+  ruleset: Ruleset;
```
Comment (mattcompiles) [inline packages/css/src/conditionalRulesets.ts:15]: `Array<Condition>`

File: packages/css/src/conditionalRulesets.ts [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+  /**
+   * Stores information about where condtions must in relation to other conditions
+   *
+   * e.g. mobile -> tablet, desktop
+   */
+  orderPriorities: Map<string, Set<String>>;
```
Comment (mattcompiles) [inline packages/css/src/conditionalRulesets.ts:21]: `precedenceLookup`

File: packages/css/src/conditionalRulesets.ts [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+  rules: Array<Rule>;
+  children: ConditionalRuleset;
+};
+
+type Ruleset = Array<ConditionInfo>;
+
+export class ConditionalRuleset {
+  ruleset: Ruleset;
+  /**
+   * Stores information about where condtions must in relation to other conditions
```
Comment (mattcompiles) [inline packages/css/src/conditionalRulesets.ts:17]:
```suggestion
   * Stores information about where conditions must in relation to other conditions
```

File: packages/css/src/conditionalRulesets.ts [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+   * e.g. mobile -> tablet, desktop
+   */
+  orderPriorities: Map<string, Set<String>>;
+
+  constructor() {
+    this.ruleset = [];
+    this.orderPriorities = new Map();
+  }
+
+  getRelevantCondition(conditions: Array<string>) {
```
Comment (mattcompiles) [inline packages/css/src/conditionalRulesets.ts:28]: `getConditionByPath(conditionPath)`

File: packages/css/src/conditionalRulesets.ts [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+    }
+
+    targetCondition.rules.push(rule);
+  }
+
+  addConditionPriorities(
+    parentConditions: Array<string>,
+    conditionOrder: Array<string>,
+  ) {
+    const targetCondition = this.getRelevantCondition(parentConditions);
```
Comment (mattcompiles) [inline packages/css/src/conditionalRulesets.ts:67]: `getConditionalRulesetByPath`

File: packages/css/src/conditionalRulesets.ts [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    const targetCondition = ruleset.findOrCreateCondition(conditionQuery);
+
+    if (!targetCondition) {
+      throw new Error('Failed to add conditional rule');
+    }
+
+    targetCondition.rules.push(rule);
+  }
+
+  addConditionPriorities(
```
Comment (markdalgleish) [inline packages/css/src/conditionalRulesets.ts:67]: Precedence?

File: packages/css/src/transformCss.ts [hunk 8]
```diff
@@ -220,29 +122,56 @@ class Stylesheet {
...
+    if (
+      !activeConditionalRuleset.mergeIfCompatible(this.currConditionalRuleset)
+    ) {
+      // Ruleset merge failed due to incompatibility. We now deopt by starting a fresh ConditionalRuleset
+      this.conditionalRulesets.push(this.currConditionalRuleset);
+    }
   }
 
-  addRule(cssRule: CSSRule) {
+  addConditionalRule(cssRule: CSSRule, conditions: Array<string>) {
```
Comment (markdalgleish) [inline packages/css/src/transformCss.ts:144]: `conditionPath`?

File: packages/css/src/conditionalRulesets.ts [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+  addConditionPrecedence(
+    conditionPath: Array<string>,
+    conditionOrder: Array<string>,
+  ) {
+    const ruleset = this.getConditionalRulesetByPath(conditionPath);
+
+    for (let i = 0; i < conditionOrder.length; i++) {
+      const condition = conditionOrder[i];
+
+      const conditionPriority =
```
Comment (markdalgleish) [inline packages/css/src/conditionalRulesets.ts:76]: Can we update all references of "priority" to "precedence"?

**Review by markdalgleish — APPROVED** [review]

**Pull request merged**